### PR TITLE
refactor(core): cleaner algo to resolve race condition between pool return and close

### LINKED ISSUE #1813: Flaky ReaderPoolTest.testGetAndCloseRace test
### Describe the bug

Seems we have a race condition in `ReaderPoolTest.testGetAndCloseRace`
One of the failures is TableReader double close error 
Another error attached below

### To reproduce

run `ReaderPoolTest.testGetAndCloseRace` for a period of time

### Expected Behavior

_No response_

### Environment

```markdown
- **QuestDB version**:
- **OS**:
- **Browser**:
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by bluestreak01**
Closes #1813

Full disclosure: the error I was able to reproduce was different. It was related to my local logger setup. This setup was rotating log files, interfering with closed file counter in the test.

On my M1 test ran 30 min without failure.

#### Extras

Reduced effort in each test to assert memory leak. Memory tags only evaluated when total usage before and after does not match. I feel instances where exact memory exchange between tags are non-existent.

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 5 / 5 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|5|5|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 5 / 5 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|5|5|100.00%|

File: core/src/main/java/io/questdb/cairo/pool/ReaderPool.java [hunk 1]
```diff
@@ -369,10 +371,7 @@ public void close() {
                 if (pool != null && entry != null && pool.returnToPool(this)) {
```
Comment (ideoma) [inline core/src/main/java/io/questdb/cairo/pool/ReaderPool.java:371]: I think `pool` has to be captured on stack. If I insert pause after `null` check but before call to `returnToPool` I get NPE.

```
                if (pool != null && entry != null) {
                    Os.pause();
                    if (pool.returnToPool(this)) {
                        return;
                    }
                }
                super.close();

```

**Review by ideoma — COMMENTED** [review]
I cannot reproduce original failure anymore. I think it can still have one theoretical flaw

File: core/src/main/java/io/questdb/cairo/pool/ReaderPool.java (hunk 1 above)
Comment (bluestreak01) [reply core/src/main/java/io/questdb/cairo/pool/ReaderPool.java]: Good spot, thank you 🙏 Fixed here: https://github.com/questdb/questdb/pull/1821/commits/ab43855402ecdfd42d14afd7e3f4a12e6e749a2c

Before the fix 2min run would produce NPE. After fix, 10min run - no issues

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 5 / 5 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|5|5|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 5 / 5 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|5|5|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 5 / 5 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|5|5|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 5 / 5 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|5|5|100.00%|

**Review by ideoma — APPROVED** [review]

**Pull request merged**